resources/models.py

```diff
@@ -536,6 +542,40 @@ def save(self, *args, **kwargs):
                 other_main_images.update(type="other")
         return super(ResourceImage, self).save(*args, **kwargs)
 
+    def full_clean(self, exclude=(), validate_unique=True):
+        if "image" not in exclude:
+            self._process_image()
+        return super(ResourceImage, self).full_clean(exclude, validate_unique)
+
+    def _process_image(self):
+        """
+        Preprocess the uploaded image file, if required.
+
+        This may transcode the image to a PNG if it's not a JPEG or PNG to begin with.
+
+        :raises InvalidImage: Exception raised if the uploaded file is not valid.
+        """
+        if not self.image:  # No image set - we can't do this right now
+            return
+
+        if self.image_format:  # Assume that if image_format is set, no further processing is required
+            return
+
+        try:
+            img = Image.open(self.image)
+            img.load()
+        except Exception as exc:
+            raise InvalidImage("Image %s not valid (%s)" % (self.image, exc)) from exc
+
+        if img.format not in ("JPEG", "PNG"):  # Needs transcoding.
+            # TODO: Maybe some heuristics about when it'd be a better idea to transcode to a lossy JPEG instead?
+            image_bio = BytesIO()
+            img.save(image_bio, format="PNG")
```

Comment:
It might be better to decide the format based on `Image.type`. Types `main` and `other` would get a JPEG format, and `map` and `ground_plan` would get converted to PNG.